From 643642187883e84b71814503089d185aff20ba6b Mon Sep 17 00:00:00 2001
From: "Vital Yep, Marco A" <marco.a.vital.yep@intel.com>
Date: Mon, 13 Jun 2016 12:08:08 -0700
Subject: [PATCH 426/639] Synopsys GMAC: Enabled TSO by default

Enabled the TSO feature by default to improve the TCP upstream
performance.

Signed-off-by: Vital Yep, Marco A <marco.a.vital.yep@intel.com>
---
 drivers/net/ethernet/synopsys/DWC_ETH_QOS_pci.c | 11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/drivers/net/ethernet/synopsys/DWC_ETH_QOS_pci.c b/drivers/net/ethernet/synopsys/DWC_ETH_QOS_pci.c
index 3c3b308..e8631df 100644
--- a/drivers/net/ethernet/synopsys/DWC_ETH_QOS_pci.c
+++ b/drivers/net/ethernet/synopsys/DWC_ETH_QOS_pci.c
@@ -50,7 +50,7 @@ static uint8_t dev_addr[6] = {0, 0x55, 0x7b, 0xb5, 0x7d, 0xf7};
 uint32_t dwc_eth_qos_pci_base_addr;
 bool config_prints = false;
 static bool msi_mode = true;
-static bool tso_enable = false;
+static bool tso_enable = true;
 /* Enabling only one queue by default because MAC sends
    most of the traffic to MTL FIFO 0. */
 static short num_of_queues = 1;
@@ -505,8 +505,13 @@ int DWC_ETH_QOS_probe(struct pci_dev *pdev,
 
    // Disable MDIO
    pdata->hw_feat.sma_sel = 0;
-   // Override TSO with module parameter
-   pdata->hw_feat.tso_en = tso_enable;
+	/* Override TSO with module parameter (if HW supports TSO) */
+	if (pdata->hw_feat.tso_en)
+		pdata->hw_feat.tso_en = tso_enable;
+	/* Notify of potential known issues with TSO in core v4.00 */
+	if (pdata->hw_feat.tso_en && pdata->version == MAC_VER_4_00 &&
+		num_of_queues > 1)
+		WRN_PRINT("TSO in v4.00 with more than one queue may fail!\n");
 
 #ifdef GBE_DEBUG
    // Force enable TSO if mss parameter was passed
-- 
2.10.1

